mail_browser_view/models/mail_mail.py

```diff
@@ -0,0 +1,106 @@
+# Copyright 2018 Camptocamp
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, models, fields
+from odoo.exceptions import MissingError
+from uuid import uuid4
+from base64 import urlsafe_b64encode, urlsafe_b64decode
+import binascii
+import lxml
+from werkzeug.urls import url_parse
+from datetime import date
+from dateutil.relativedelta import relativedelta
+
+
+class Mail(models.Model):
+    _inherit = 'mail.mail'
+
+    @api.model
+    def create(self, vals):
+        rec = super(Mail, self).create(vals)
+        rec._replace_view_url()
+        return rec
+
+    @api.model
+    def get_record_for_token(self, token):
+        try:
+            token = urlsafe_b64decode(token).decode()
+            access_token, rec_id = token[:32], token[32:]
+            rec = self.sudo().browse(int(rec_id))
+            return (rec.access_token == access_token and
+                    rec.is_token_alive and rec or self)
+        except (ValueError, MissingError, binascii.Error):
+            return self
```

Comment:
I'd return None or self.browse() as none value since you might call this from a recordset and you'll get the same record back.